tools/build/compute_version.go

```diff
@@ -0,0 +1,70 @@
+// Package main is a tool to compute the version of the project.
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <output-file>\n", os.Args[0])
+		os.Exit(2)
+	}
+
+	f, err := os.Create(os.Args[1])
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+	defer f.Close()
+
+	tag, err := getTag()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+
+	if dirty, err := isDirty(); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	} else if dirty {
+		tag += "-dirty"
+	}
+
+	fmt.Print(tag)
+	fmt.Fprintln(f, tag)
+}
+
+func getTag() (string, error) {
+	out, err := exec.Command("git", "describe", "HEAD").CombinedOutput()
+	if err == nil {
+		return strings.TrimSpace(string(out)), nil
+	}
+
+	fmt.Fprintf(os.Stderr, "Warning: using default tagname because we could not get tag: git describe HEAD: %v. %s", err, out)
+
+	out, err = exec.Command("git", "rev-parse", "HEAD").CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("git rev-parse HEAD: %v. %s", err, out)
+	}
+	commit := strings.TrimSpace(string(out[:8]))
+
+	// Start with 0 so that it is always less than any tag.
+	return fmt.Sprintf("0-untagged-%s", commit), nil
+}
+
+func isDirty() (bool, error) {
+	out, err := exec.Command("git", "status").Output()
```

Comment:
`git status --porcelain`  outputs nothing if the working tree is clean, eliminating the need to verify the contents of the output if you just want to know that.